.dotnet/src/Custom/Batch/BatchOperation.Protocol.cs

```diff
@@ -0,0 +1,287 @@
+
+        await WaitForCompletionAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override void WaitForCompletion(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            _pollingInterval.Wait(response, cancellationToken);
+
+            ClientResult result = GetBatch(_batchId, cancellationToken.ToRequestOptions());
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public void WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        WaitForCompletion(cancellationToken);
+    }
+
+    private void ApplyUpdate(ClientResult result)
+    {
+        PipelineResponse response = result.GetRawResponse();
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string? status = doc.RootElement.GetProperty("status"u8).GetString();
+
+        IsCompleted = GetIsCompleted(status);
+        SetRawResponse(response);
+    }
+
+    private static bool GetIsCompleted(string? status)
+    {
+        return status == InternalBatchStatus.Completed.ToString() ||
+            status == InternalBatchStatus.Cancelled.ToString() ||
+            status == InternalBatchStatus.Expired.ToString() ||
+            status == InternalBatchStatus.Failed.ToString();
+    }
+
+    // Generated protocol methods
+
+    /// <summary>
+    /// [Protocol Method] Retrieves a batch.
+    /// </summary>
+    /// <param name="batchId"> The ID of the batch to retrieve. </param>
+    /// <param name="options"> The request options, which can override default behaviors of the client pipeline on a per-call basis. </param>
+    /// <exception cref="ArgumentNullException"> <paramref name="batchId"/> is null. </exception>
+    /// <exception cref="ArgumentException"> <paramref name="batchId"/> is an empty string, and was expected to be non-empty. </exception>
+    /// <exception cref="ClientResultException"> Service returned a non-success status code. </exception>
+    /// <returns> The response returned from the service. </returns>
+    public virtual async Task<ClientResult> GetBatchAsync(string batchId, RequestOptions options)
+    {
+        Argument.AssertNotNullOrEmpty(batchId, nameof(batchId));
+
+        using PipelineMessage message = CreateRetrieveBatchRequest(batchId, options);
+        return ClientResult.FromResponse(await _pipeline.ProcessMessageAsync(message, options).ConfigureAwait(false));
```

Comment:
it would be good if we had a helper for this line of code. i.e.

return await _pipeline.GetResultAsync(message,options).ConfigureAwait(false);

Reply:
I think this is also generated code -- would it be OK if we followed up on this one in a separate PR?

Reply:
Saving this to come back to in a separate PR